File: elixir/easy/0070/0070.exs
defmodule Solution do
  @spec climb_stairs(n :: integer) :: integer
  def climb_stairs(n) do
    _climb_stairs(n, 1, 0, 1)
  end

  defp _climb_stairs(n, n, a, b), do: a + b
  defp _climb_stairs(n, step, a, b), do: _climb_stairs(n, step + 1, b, a + b)
end

ExUnit.start()

defmodule AssertionTest do
  use ExUnit.Case, async: true

  test "test_1" do
    assert Solution.climb_stairs(3) == 3
  end

  test "test_2" do
    assert Solution.climb_stairs(4) == 5
  end
end

File: py/easy/0070/0070.py
from functools import cache


class Solution:
    def climbStairs(self, n: int) -> int:
        @cache
        def _climb_stairs(n):
            if n <= 2:
                return n
            return _climb_stairs(n - 1) + _climb_stairs(n - 2)

        return _climb_stairs(n)


solution = Solution()
assert solution.climbStairs(1) == 1
assert solution.climbStairs(2) == 2
assert solution.climbStairs(3) == 3
assert solution.climbStairs(4) == 5
